lib/inventory/store.go

```diff
@@ -0,0 +1,184 @@
+	// shards is an array of individually locked mappings of the form serverID -> handle(s).
+	// keys are assigned to shards via a maphash.
+	shards [SHARDS]*shard
+	// seed is used to deterministically calculate key hashes in order to select
+	// the correct shard for a given key.
+	seed maphash.Seed
+}
+
+// NewStore creates a new inventory control handle store.
+func NewStore() *Store {
+	var shards [SHARDS]*shard
+	for i := range shards {
+		shards[i] = newShard()
+	}
+	return &Store{
+		shards: shards,
+		seed:   maphash.MakeSeed(),
+	}
+}
+
+// Get attempts to load a handle for the given server ID.
+// note: if multiple handles exist for a given server, the returned handle
+// is selected pseudorandomly from the available set.
+func (s *Store) Get(serverID string) (handle UpstreamHandle, ok bool) {
+	return s.getShard(serverID).get(serverID)
+}
+
+// Insert adds a new handle to the store.
+func (s *Store) Insert(handle UpstreamHandle) {
+	s.getShard(handle.Hello().ServerID).insert(handle)
+}
+
+// Remove removes the handle from the store.
+func (s *Store) Remove(handle UpstreamHandle) {
+	s.getShard(handle.Hello().ServerID).remove(handle)
+}
+
+// Iter iterates across all handles registered with this store.
+// note: if multiple handles are registered for a given server, only
+// one handle is selected pseudorandomly to be observed.
+func (s *Store) Iter(fn func(UpstreamHandle)) {
+	for _, shard := range s.shards {
+		shard.iter(fn)
+	}
+}
+
+// Len returns the count of currently registered servers (servers with
+// multiple handles registered still only count as one).
+func (s *Store) Len() int {
+	var total int
+	for _, shard := range s.shards {
+		total += shard.Len()
+	}
+	return total
+}
+
+// getShard loads the shard for the given serverID.
+func (s *Store) getShard(serverID string) *shard {
+	var h maphash.Hash
+	// all hashes must use the same seed in order for subsequent calls
+	// to land at the same shard for a given serverID.
+	h.SetSeed(s.seed)
+	h.WriteString(serverID)
+	idx := h.Sum64() % uint64(SHARDS)
+	return s.shards[int(idx)]
+}
+
+type shard struct {
+	// rw protects inner mapping
+	rw sync.RWMutex
+	// mapping of server id => handle(s).
+	m map[string]*entry
+}
+
+func newShard() *shard {
+	return &shard{
+		m: make(map[string]*entry),
+	}
+}
+
+type entry struct {
+	// ct is atomically incremented as a means to pseudorandomly distribute
+	// load for instances that have multiple handles registered.
+	ct      *atomic.Uint64
+	handles []UpstreamHandle
+}
+
+func (s *shard) get(serverID string) (handle UpstreamHandle, ok bool) {
+	s.rw.RLock()
+	defer s.rw.RUnlock()
+	entry, ok := s.m[serverID]
+	if !ok {
+		return nil, false
+	}
+	idx := entry.ct.Inc() % uint64(len(entry.handles))
+	handle = entry.handles[int(idx)]
+	return handle, true
+}
+
+func (s *shard) iter(fn func(UpstreamHandle)) {
+	s.rw.RLock()
+	defer s.rw.RUnlock()
+	for _, entry := range s.m {
+		idx := entry.ct.Inc() % uint64(len(entry.handles))
```

Comment:
I believe this code is at risk of a divide by zero panic.

Reply:
Yeah, we should either have length 0 as a special case or we should never have a zero-length `handles` by removing the whole entry when the last handle is removed.

Reply:
Yeah, messed up my removal logic at some point.  Fixed panic and updated test to cover this path.

Reply:
@fspmarshall please make sure this fix makes it in to master.

Great catch @strideynet!